### PR TITLE
add failing tests for all the helper APIs

### PR DESCRIPTION
**Pull request by fivetanley**
cc @stefanpenner @mixonic

Let's solicit feedback here before implementing.

### CODE REVIEW

**Comment by rwjblue** [comment]
Can you add tests for:

* Usage as an inline `{{if}}` (i.e. `<div class="foo {{if (await promise) 'fullfilled' 'rejected'}}"></div>`)
* Usage with `{{each}}` (i.e. `{{#each (await promise) as |item|}}`)

**Comment by fivetanley** [comment]
Does writing the helper change for those scenarios? I'm happy to add more tests, but at that point it feels like testing the framework.

**Comment by fivetanley** [comment]
Although, in a magical world I would like to build, library authors could subscribe to ember updates and run ember-try. /me dreams

**Comment by rwjblue** [comment]
@fivetanley - Recomputation is actually somewhat complicated.  It seems reasonable to me that we should have tests for the primary use cases that we expect folks to use this for (even if just to know when they stop working)...

**Comment by fivetanley** [comment]
tests added https://github.com/fivetanley/ember-promise-helpers/commit/f6474183f2f79d84eca2c25c720eecd61c1625ee

**Comment by stefanpenner** [comment]
Some additional tests would be the promise changing overtime. Unfortunately without cancellation, we can't totally cleanup, but we should prevent an earlier promises's settling of affecting the outcome if it is no longer the active promise.

**Comment by stefanpenner** [comment]
Additionally, we may want to consider when rejections should hit `.on('error')` and when they shouldn't. At-least have a game-plan for it.

**Comment by fivetanley** [comment]
> Additionally, we may want to consider when rejections should hit .on('error') and when they shouldn't. At-least have a game-plan for it.

I am more keen to say they should always hit `.on('error')`, from a debuggability perspective

**Comment by stefanpenner** [comment]
> I am more keen to say they should always hit .on('error'), from a debuggability perspective

I think it needs some extra thought (tricky to strike the right balance). For example, what if someone implements `(rejected promise)` should we still fail a test? Should we every fail a test? Do we fail tests for specific failures? Do we provide a (error-handled) helper?

**Comment by fivetanley** [comment]
> For example, what if someone implements (rejected promise) should we still fail a test?

Is it possible for us to tell the difference from userland?

**Comment by stefanpenner** [comment]
> Is it possible for us to tell the difference from userland?

ya

**Comment by fivetanley** [comment]
Care to elaborate? :p It seems like all promises would respond to `catch/then`.

**Comment by tchak** [comment]
Not a huge fan of `promise-rejected-reason` name. Maybe something like `await-rejection`?
I also liked the block version of `await` in the RFC : `{{#await model.author as |author, error, isPending|}}`. I think both version can be useful.

**Comment by fivetanley** [comment]
@stefanpenner Added tests for timing stuff 8f48fec

File: tests/integration/is-pending-test.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  integration: true
+});
+
+test('is true until the promise resolves', function (assert) {
+  let deferred = RSVP.defer();
+
+  this.set('promise', deferred.promise);
+
+  this.render(hbs`
+    {{#if (is-pending)}}
```
Comment (mixonic) [inline tests/integration/is-pending-test.js]: Must be passed `promise`

**Comment by mixonic** [comment]
I'm confused as to if `await` is intended to return the resolved value or just a boolean. IMO `await` should return the resolved content and `is-resolved` could return a boolean mirroring `is-rejected`.

**Comment by fivetanley** [comment]
> IMO await should return the resolved content

That is the intended behavior, I will update the tests to reflect that.

**Pull request merged**